backend/de.metas.contracts/src/main/java/de/metas/contracts/callorder/detail/CallOrderDetailRepo.java

```diff
@@ -0,0 +1,244 @@
+
+		final ICompositeQueryUpdater<I_C_CallOrderDetail> updater = queryBL
+				.createCompositeQueryUpdater(I_C_CallOrderDetail.class)
+				.addSetColumnValue(I_C_CallOrderDetail.COLUMNNAME_QtyDeliveredInUOM, BigDecimal.ZERO);
+
+		detailsMatchingShipmentQuery.update(updater);
+	}
+
+	public void resetInvoicedQtyForInvoice(@NonNull final InvoiceId invoiceId)
+	{
+		final IQuery<I_C_CallOrderDetail> detailsMatchingInvoiceQuery = queryBL.createQueryBuilder(I_C_CallOrderDetail.class)
+				.addEqualsFilter(I_C_CallOrderDetail.COLUMNNAME_C_Invoice_ID, invoiceId)
+				.create();
+
+		final ICompositeQueryUpdater<I_C_CallOrderDetail> updater = queryBL
+				.createCompositeQueryUpdater(I_C_CallOrderDetail.class)
+				.addSetColumnValue(I_C_CallOrderDetail.COLUMNNAME_QtyInvoicedInUOM, BigDecimal.ZERO);
+
+		detailsMatchingInvoiceQuery.update(updater);
+	}
+
+	public void resetEnteredQtyForOrder(@NonNull final OrderId orderId)
+	{
+		final IQuery<I_C_CallOrderDetail> detailsMatchingOrderQuery = queryBL.createQueryBuilder(I_C_CallOrderDetail.class)
+				.addEqualsFilter(I_C_CallOrderDetail.COLUMNNAME_C_Order_ID, orderId)
+				.create();
+
+		final ICompositeQueryUpdater<I_C_CallOrderDetail> resetQtyEntered = queryBL
+				.createCompositeQueryUpdater(I_C_CallOrderDetail.class)
+				.addSetColumnValue(I_C_CallOrderDetail.COLUMNNAME_QtyEntered, BigDecimal.ZERO);
+
+		detailsMatchingOrderQuery
+				.update(resetQtyEntered);
+	}
+
+	@NonNull
+	public Optional<CallOrderDetail> getDetailByQuery(@NonNull final CallOrderDetailQuery query)
+	{
+		final IQueryBuilder<I_C_CallOrderDetail> queryBuilder = queryBL.createQueryBuilder(I_C_CallOrderDetail.class)
+				.addEqualsFilter(I_C_CallOrderDetail.COLUMNNAME_C_CallOrderSummary_ID, query.getSummaryId());
+
+		if (query.getOrderLineId() != null)
+		{
+			queryBuilder.addEqualsFilter(I_C_CallOrderDetail.COLUMNNAME_C_OrderLine_ID, query.getOrderLineId());
+		}
+
+		if (query.getInOutLineId() != null)
+		{
+			queryBuilder.addEqualsFilter(I_C_CallOrderDetail.COLUMNNAME_M_InOutLine_ID, query.getInOutLineId());
+		}
+
+		if (query.getInvoiceLineId() != null)
+		{
+			queryBuilder.addEqualsFilter(I_C_CallOrderDetail.COLUMN_C_InvoiceLine_ID, query.getInvoiceLineId());
+		}
+
+		return queryBuilder.create()
+				.firstOnlyOptional(I_C_CallOrderDetail.class)
+				.map(CallOrderDetailRepo::ofRecord);
+	}
+
+	private static void setDetailDataToRecord(@NonNull final I_C_CallOrderDetail record, @NonNull final CallOrderDetailData detailData)
+	{
+		record.setC_CallOrderSummary_ID(detailData.getSummaryId().getRepoId());
+
+		if (detailData.getOrderDetail() != null)
+		{
+			final CallOrderDetailData.OrderDetail orderDetail = detailData.getOrderDetail();
+
+			record.setC_OrderLine_ID(orderDetail.getOrderLineId().getRepoId());
+			record.setC_Order_ID(orderDetail.getOrderId().getRepoId());
+			record.setQtyEntered(orderDetail.getQtyEntered().toBigDecimal());
+			record.setC_UOM_ID(orderDetail.getQtyEntered().getUomId().getRepoId());
```

Comment:
i think we have to make sure that all qtys (entered, deliveredInUOM etc) are really converted to the same UOM. might be that the order-line and the inout-line have different UOMs

Reply:
yeah, I know what you mean, but there can't be a `c_callorderdetail` record with both c_orderline and m_inoutline set... we create one individual detail record for each step.. i.e.:

- let's say we create and complete a call order with one line, after complete we will get a c_callorderdetail record referencing the c_orderline of the call order
- then, when we ship that order, we will get a `NEW` `c_callorderdetail` record referencing the shipment line... so at this point for the same call order we have 2 `c_callorderdetail` records, one for the order line and one for the shipment line
- and then, when we create the invoice, there will be another `NEW` `c_callorderdetail` referencing the `c_invoiceline`